### PR TITLE
Create base Daemon class and refactor Undertaker, Reaper, Abacus Account and Judge Repairer daemons to inherit from base class

### PR DESCRIPTION
**Pull request by rdimaio**
Initial commit where base class Daemon is created, and Undertaker inherits from Daemon. If approved, I will refactor the other daemons in the same way. Tested in local fork.

(note: the changes in `common.py` outside of the main `Daemon` class are due to the `black` formatter adjusting style to meet PEP8)

Initial commit for #6478

### CODE REVIEW

**Review by rcarpa — DISMISSED** [review]

**Comment by rdimaio** [comment]
In https://github.com/rucio/rucio/pull/6489/commits/29a5666a6ccb39332cf4178117217b8b29c55ebf, Reaper is also refactored to follow the base Daemon class. Tests pass on fork branch: https://github.com/rdimaio/rucio/commit/29a5666a6ccb39332cf4178117217b8b29c55ebf

File: lib/rucio/daemons/common.py
Comment (dchristidis) [inline lib/rucio/daemons/common.py]: Please sort modules and imported functions/classes alphabetically. This may apply elsewhere as well.

File: lib/rucio/daemons/common.py
Comment (dchristidis) [inline lib/rucio/daemons/common.py]: I see no need for the hints to be quoted.

File: lib/rucio/daemons/common.py
Comment (dchristidis) [inline lib/rucio/daemons/common.py]: I wonder if it’s useful to use `typing.Final` here.

File: lib/rucio/daemons/common.py
Comment (dchristidis) [inline lib/rucio/daemons/common.py]: It is not recommended to do string interpolation outside and prior to a `logging.log` call. See: https://docs.python.org/3/howto/logging.html#optimization

This may apply elsewhere as well.

File: lib/rucio/daemons/common.py
Comment (dchristidis) [inline lib/rucio/daemons/common.py]: I never particularly liked this. Pylint should complain with `expression-not-assigned`. Is it maybe a good opportunity to split into two lines?

File: lib/rucio/daemons/common.py
Comment (dchristidis) [inline lib/rucio/daemons/common.py]: Is there any usefulness to specifying `start=0`, which is already the default?

File: lib/rucio/daemons/common.py
Comment (dchristidis) [inline lib/rucio/daemons/common.py]: Firstly, I think placing the comment above the block makes it unclear. Better put it together with line 129.

But more importantly, do you understand this loop? `threading.Thread.join()` always returns `None`, according to its documentation. So under what circumstances could we have multiple iterations?

File: lib/rucio/daemons/common.py
Comment (dchristidis) [inline lib/rucio/daemons/common.py]: Why is `if renewal_interval` necessary here? The hint is `int`.

File: lib/rucio/daemons/common.py
Comment (dchristidis) [inline lib/rucio/daemons/common.py]: I don’t think that I can agree that this improves readability.

File: lib/rucio/daemons/common.py
Comment (dchristidis) [inline lib/rucio/daemons/common.py:92]: I wonder if we should take the opportunity to rethink some of the chosen names. There’s ‘pre-run’, ‘run’, ‘run once’, ‘call daemon’. Do they feel intuitive? Had I not known, I would have guessed that `_call_daemon()` is the outer function, `_pre_run_checks()` is executed before `run()`, and that `run()` and `_run_once()` are alternatives.

**Review by dchristidis — CHANGES_REQUESTED** [review]
My initial impression was positive; I think the proposed design has a lot of merit. But I would not recommend its inclusion in its current state.

What hinders the review is the mix of functional and aesthetical changes in the same commit. Things like folding long lines, changing indentation levels, and adding type hints to existing functions, are all very welcome, but bundling them together with the actual behavioural changes for PRs as large as this should be discouraged.

Then there’s the change of existing strings to use double instead of single quotes. The Rucio project does not have an official guideline, so the recommendation is to follow the existing pattern when modifying code and using one’s preference when introducing big blocks of new code. But I’m not too enthusiastic about trying to conform an entire existing file. What if the next developer that works on it thinks that single quotes are better? It can also be an annoyance when going through the history of a file.

I stopped halfway through `lib/rucio/daemons/common.py`, without looking too close at the rest. I would be very happy to continue the review, but if ends up being up to me, I think the PR may need to go through multiple iterations before I can approve it.

PS: Afterwords I saw that:

> (note: the changes in common.py outside of the main Daemon class are due to the black formatter adjusting style to meet PEP8)

Black is highly opinionated. Its use induced negative reactions in the past. I would be careful.

**Comment by rdimaio** [comment]
> Things like folding long lines, changing indentation levels, and adding type hints to existing functions, are all very welcome, but bundling them together with the actual behavioural changes for PRs as large as this should be discouraged.

I agree on all points except for type hinting. On the formatting side, I recognize I made a mistake in using a formatter, as it makes it harder to spot the functional changes; I'm happy to do whatever you think it's best to do here, e.g. if you want me to amend the commit so that the formatting changes are removed I'm okay with doing that, but if you have other ideas let me know. I will be careful about this for future PRs.

For the type hinting to existing functions: the functions whose type hints have been changed/added in this PR are related to the daemons that have been refactored (e.g. `parse_expression` is used in Reaper). I added type hints to those functions while refactoring the daemons in order to better understand the logic in the daemons themselves. I understand that these type hints are not part of the "refactoring" itself, but I personally don't see an issue in having them within the same PR, but of course this is just my opinion so I'm okay with separating them if you prefer.

File: lib/rucio/daemons/common.py
Comment (rdimaio) [reply lib/rucio/daemons/common.py]: To avoid doing this kind of work manually (which would be a waste of time), would it make sense to enforce something like `isort` as part of the pre-commit hook? https://pycqa.github.io/isort/

**Comment by dchristidis** [comment]
It’s the scale of the PR that makes this matter to me. If it were a few-line change in a single function, that also happened to add type hints, then I would not object to a single commit. This, however, is over one thousand lines and spans multiple files. Hence why I tend to request that behavioural and non-behavioural changes are separated (and type hints are part of the latter). Both are welcome, of course! But as a reviewer, it aids me to know when a commit is not supposed to make any effective changes.

So yes, if Martin expects the final say from me, then I need to ask you to rewrite the commits as described.

File: lib/rucio/daemons/common.py
Comment (dchristidis) [reply lib/rucio/daemons/common.py]: That’s an interesting proposal. I would recommend opening a separate issue to discuss this and to consider applying it across the entire codebase beforehand.

Comment (rdimaio) [reply lib/rucio/daemons/common.py]: I had opened this some weeks ago https://github.com/rucio/rucio/issues/6488 where I mentioned `isort` as well - I will bring it up in the retreat and see what the overall thoughts are. For now, I will make this change manually

File: lib/rucio/daemons/common.py
Comment (rdimaio) [reply lib/rucio/daemons/common.py]: Good idea, it's definitely appropriate here, will add it now

File: lib/rucio/daemons/common.py
Comment (rdimaio) [reply lib/rucio/daemons/common.py:92]: I am okay with changing the names if there are any suggestions, but I could not think of anything else at the moment. There are some things that I do not like, e.g. I think `once` and consequently `run_once` are not clear enough, but those names currently match some of the names from the daemon's user's perspective (e.g. the `--run-once` flag) so I didn't want to change them

File: lib/rucio/daemons/common.py
Comment (rdimaio) [reply lib/rucio/daemons/common.py]: I think the reference recommends not to interpolate strings when that involves running a potentially expensive function, and accessing an instance attribute would likely not be expensive. 

Is there any alternative to doing this if I want to log the daemon name? Alternatively we can just log "daemon: starting threads" without specifying the daemon name, but that might be confusing if there are multiple daemons running at the same time

File: lib/rucio/daemons/common.py
Comment (rdimaio) [reply lib/rucio/daemons/common.py]: That's how it was written in `Undertaker` and I didn't notice it while refactoring. Removed `0`

File: lib/rucio/daemons/common.py
Comment (rdimaio) [reply lib/rucio/daemons/common.py]: This whole code block was copied from Reaper:

```
    # Interruptible joins require a timeout.
    while threads_list:
        threads_list = [thread.join(timeout=3.14) for thread in threads_list if thread and thread.is_alive()]
```

In general, every daemon involved in the refactoring had a loop of this kind, e.g. in Undertaker:

```
       # Interruptible joins require a timeout.
        while threads[0].is_alive():
            [t.join(timeout=3.14) for t in threads]
```

Based on the fact that `threading.Thread.join()` returns `None`, then I think the version of the loop present in Undertaker is the one that's actually correct, given that it loops as long as the first thread is alive, right?

File: lib/rucio/daemons/common.py
Comment (rdimaio) [reply lib/rucio/daemons/common.py]: That's the way it was written in the refactored daemons, e.g. in Judge Repairer:

```
def stop(signum: "Optional[int]" = None, frame: "Optional[FrameType]" = None) -> None:
    """
    Graceful exit.
    """

    graceful_stop.set()
```

I agree that the quotes are not necessary here, removed.

File: lib/rucio/daemons/common.py
Comment (dchristidis) [reply lib/rucio/daemons/common.py]: You don’t have to avoid interpolation altogether, just *defer* it to `logging`. See Pylint’s `logging-fstring-interpolation`. The solution is:

```python
logging.info("%s: starting threads", daemon_name)
```

**Comment by rdimaio** [comment]
> It’s the scale of the PR that makes this matter to me. If it were a few-line change in a single function, that also happened to add type hints, then I would not object to a single commit. This, however, is over one thousand lines and spans multiple files. Hence why I tend to request that behavioural and non-behavioural changes are separated (and type hints are part of the latter). Both are welcome, of course! But as a reviewer, it aids me to know when a commit is not supposed to make any effective changes.
> 
> So yes, if Martin expects the final say from me, then I need to ask you to rewrite the commits as described.

I've rebased the PR to remove all unnecessary non-behavioural changes

File: lib/rucio/daemons/common.py
Comment (rdimaio) [reply lib/rucio/daemons/common.py]: Thanks, amended changes to use this instead (also changed other places where I was logging with fstring)

File: lib/rucio/daemons/common.py
Comment (dchristidis) [reply lib/rucio/daemons/common.py]: I’m afraid I don’t know. That’s why I’m genuinely asking, because you’re (rightfully) copying existing implementations, and maybe they copied it over from somewhere else, but I wonder if anyone really understands them.

So, since the long-term goal of your effort is to implement this exactly once, I am taking the opportunity to try to understand exactly what it is supposed to do.

Comment (dchristidis) [reply lib/rucio/daemons/common.py]: The more I look at the existing implementations, the more perplexed I am. Won’t the loop in the Reaper exit after one iteration, while the threads are still running?

```python
>>> import threading
>>> import time
>>> def worker():
...     while True:
...         time.sleep(60)
... 
>>> threads_list = [threading.Thread(target=worker) for _ in range(2)]
>>> for thread in threads_list:
...     thread.start()
... 
>>> while threads_list:
...     print(threads_list)
...     threads_list = [thread.join(timeout=0.1) for thread in threads_list if thread and thread.is_alive()]
... 
[<Thread(Thread-1 (worker), started 140304535520960)>, <Thread(Thread-2 (worker), started 140304525035200)>]
[None, None]
>>> # Huh?
```

Comment (dchristidis) [reply lib/rucio/daemons/common.py]: Following my own little tests, I am starting to question the need for a timeout and whether the try-except structure we have have in the daemons for keyboard interrupts is working how it is supposed to.

I am also looking for reproducible example as to why the following won’t do:

```python
for thread in threads:
    thread.join()
```

Comment (dchristidis) [reply lib/rucio/daemons/common.py]: I’m getting different results in Python 3.6 and 3.9. :shrug: I think this further drives the point that we implemented a solution in Python 2, based on the situation at the time, and kept it as it is without checking if it still makes sense.

Comment (rdimaio) [reply lib/rucio/daemons/common.py]: Thanks for looking into this! I'm not sure what's the best way forward here - the different results in different Python versions are a bit worrying (though AFAIK Rucio requires Python 3.9 to run, so at least from the user's perspective it will be consistent), but it might also be one of those things that "if it works now, better not touch it" (even though it doesn't really work).

It's definitely a carryover from Python 2 in some way, I found [this commit](https://github.com/rucio/rucio/commit/9fa89245fe468b3ce9b67cfa9dc06bac4c1b77e6) under [this issue](https://github.com/rucio/rucio/issues/1924) for Python 3 compatibility with daemons, and that logic was left untouched

Comment (dchristidis) [reply lib/rucio/daemons/common.py]: As far as I can tell, the implementation seen in the Undertaker has been there since the very beginning (over a decade ago, see b896d4c64bec067f2f3a48971ac4fa46229a5b99). The justification is lost in time.

Here’s a modern test case:

```python
#!/usr/bin/env python3

import os
import signal
import threading
import time
from types import FrameType
from typing import Optional

GRACEFUL_STOP = threading.Event()


def worker() -> None:
    while not GRACEFUL_STOP.is_set():
        print('worker active')
        time.sleep(1)
    print('worker terminating')
    time.sleep(5)
    print('worker terminated')


def stop(signum: Optional[int] = None, frame: Optional[FrameType] = None) -> None:
    GRACEFUL_STOP.set()
    if signum == signal.SIGINT:
        if hasattr(stop, 'received_sigint'):
            exit(1)
        else:
            stop.received_sigint = True

if __name__ == '__main__':
    if 'CAPTURE_SIGINT' in os.environ:
        signal.signal(signal.SIGINT, stop)
    threads = [threading.Thread(target=worker) for _ in range(1)]
    for thread in threads:
        thread.start()
    for thread in threads:
        thread.join()
    print('main thread terminated')
```

Here’s a sample run with Python 3.9:

```console
$ python3 test.py 
worker active
worker active
^CTraceback (most recent call last):
  File "/afs/cern.ch/user/d/dichrist/test.py", line 37, in <module>
    thread.join()
  File "/usr/lib64/python3.9/threading.py", line 1060, in join
    self._wait_for_tstate_lock()
  File "/usr/lib64/python3.9/threading.py", line 1080, in _wait_for_tstate_lock
    if lock.acquire(block, timeout):
KeyboardInterrupt

$ CAPTURE_SIGINT=TRUE python3 test.py 
worker active
worker active
^Cworker terminating
worker terminated
main thread terminated
$ CAPTURE_SIGINT=TRUE python3 test.py 
worker active
worker active
^C^C$ 
```

It all comes down to this:

 1. If we don’t care about graceful shutdown with `SIGINT`, then the simple implementation works.
 2. If we care about it, then we can easily adjust the `stop()` function which is already used for `SIGTERM` to exit gracefully upon the first keyboard interrupt and immediately upon the second.

@bari12 @mlassnig Thoughts?

Comment (bari12) [reply lib/rucio/daemons/common.py]: I would adjust this. We should take this as an opportunity to renew the code here and don't carry outdated (not to say wrong) choices further in the future.

Comment (dchristidis) [reply lib/rucio/daemons/common.py]: But the question remains: should we implement Ctrl+C leading to a graceful stop?

File: lib/rucio/daemons/common.py
Comment (rdimaio) [inline lib/rucio/daemons/common.py:145]: should be sigint only, not sigterm as well

File: bin/rucio-undertaker
Comment (rdimaio) [inline bin/rucio-undertaker:77]: this try except is not needed

File: lib/rucio/daemons/common.py
Comment (rdimaio) [inline lib/rucio/daemons/common.py:112]: could be renamed to start

File: lib/rucio/daemons/common.py
Comment (rdimaio) [reply lib/rucio/daemons/common.py:145]: worth noting: the daemons in their current version use `SIGTERM`, not `SIGINT`, e.g. https://github.com/rucio/rucio/blob/c8316a595af9ef6336a131ead947414f32e7a8ca/bin/rucio-auditor#L86

Comment (dchristidis) [reply lib/rucio/daemons/common.py:145]: The Auditor is not a good example.

`SIGTERM` is what the OS/supervisor/scheduler is using to terminated a process under its control. It is supposed to send `SIGKILL` after a timeout.

`SIGINT` is what Ctrl+C sends. It only makes sense when one starts a daemon process from their terminal, which is unlikely, especially in a production environment.

The daemons that are (supposedly) handling `SIGNIT` are doing so via the `KeyboardInterrupt` exception.

Comment (rdimaio) [reply lib/rucio/daemons/common.py:145]: I linked to Auditor as a random example, but it was part of this search https://github.com/search?q=repo%3Arucio%2Frucio%20sigterm&type=code where it seems like most daemons use `SIGTERM`. I agree that for the purpose of gracefully stopping manually we should use `SIGINT`, I just wanted to bring this up for reference

**Comment by rdimaio** [comment]
Closing this; we can keep the https://github.com/rucio/rucio/issues/6478 issue open, and work on this again later down the line, as it's not urgent now

**Pull request closed without merging**
